Add C++ nn::functional pad

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26605 C++ nn::ConstantPad{1,2,3}d
* #26604 C++ nn::ZeroPad2d
* #26603 C++ nn::ReplicationPad{1,2,3}d
* #26602 C++ nn::ReflectionPad1d and nn::ReflectionPad2d
* **#26601 Add C++ nn::functional pad**

Differential Revision: [D17517468](https://our.internmc.facebook.com/intern/diff/D17517468)

Commit:
Add C++ nn::functional pad

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

yf225 commented:
Comment from @ShahriarSS :

> @yf225 There are some major problems here. Let me address those and then I can review the PR in detail. 
> 1. I think that we should have test cases to check the validity of ```TORCH_CHECK```s.
> 2. I think that we shouldn't use string for the mode in options. We should use enum classes.
> 3. I've had problems with using ```IntArrayRef``` as class members. Can we replace it with something else?

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

pbelevich commented on test/cpp/api/functional.cpp:
```diff
@@ -1112,3 +1112,88 @@ TEST_F(FunctionalTest, Threshold) {
+TEST_F(FunctionalTest, Pad) {
+  {
+    auto input = torch::arange(6, torch::kDouble).reshape({1, 2, 3});
+    auto output = F::pad(input, PadOptions({1, 2}).mode(torch::kCircular));
```
Correct me if I'm wrong but this test does not cover cases `if (padding.size() > 2)` and `if (padding.size() > 4)` in `_pad_circular ` method, which have significant logic. I think we should test those cases.

yf225 replied:
I added the tests

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head

Commit:
Update on "Add C++ nn::functional pad"

Add C++ nn::functional pad

This PR adds nn::functional::pad to C++ API

gh-metadata: pytorch pytorch 26601 gh/yf225/1/head
